app/static/ringing_room.js

```diff
@@ -787,6 +811,489 @@ $(document).ready(function() {
+                this.update_comp_suggestions(next_value);
+            }
+        },
+
+        methods: {
+            /* METHODS CALLED WHEN THE USER CHANGES SOME CONTROLS */
+            on_change_sensitivity: function () {
+                socketio.emit('c_wheatley_setting', {
+                    tower_id: cur_tower_id,
+                    settings: {
+                        sensitivity: this.sensitivity
+                    }
+                });
+            },
+            on_change_use_up_down_in: function () {
+                socketio.emit('c_wheatley_setting', {
+                    tower_id: cur_tower_id,
+                    settings: {
+                        use_up_down_in: this.use_up_down_in
+                    }
+                });
+            },
+            on_change_stop_at_rounds: function () {
+                socketio.emit('c_wheatley_setting', {
+                    tower_id: cur_tower_id,
+                    settings: {
+                        stop_at_rounds: this.stop_at_rounds
+                    }
+                });
+            },
+            reset_wheatley: function () {
+                socketio.emit('c_reset_wheatley', {tower_id: cur_tower_id});
+            },
+
+            /* CALLBACKS CALLED FROM RECEIVING A SOCKETIO SIGNAL */
+            update_settings: function (new_settings) {
+                for (const key in new_settings) {
+                    if (key == 'sensitivity') {
+                        this.sensitivity = new_settings[key];
+                    }
+                    if (key == 'use_up_down_in') {
+                        this.use_up_down_in = new_settings[key];
+                    }
+                    if (key == 'stop_at_rounds') {
+                        this.stop_at_rounds = new_settings[key];
+                    }
+                }
+            },
+            update_row_gen: function (new_row_gen) {
+                if (new_row_gen) {
+                    this.row_gen = new_row_gen;
+                }
+            },
+            update_is_ringing: function (new_value) {
+                this.is_ringing = new_value;
+            },
+            update_number_of_bells: function () {
+                this.update_method_suggestions(this.method_name);
+                this.update_comp_suggestions(this.complib_id);
+            },
+
+            /* METHODS RELATED TO THE USER UPDATING THE ROW_GEN CONTROLS */
+            on_stop_touch: function () {
+                socketio.emit("c_wheatley_stop_touch", {tower_id: window.tower_parameters.id});
+            },
+            update_method_suggestions: function (partial_method_name) {
+                // Store a reference to 'this' (the vue model) as a local variable, so that it can
+                // be used in the JSON get callback to set the autocomplete results.
+                var _this = this;
+
+                if (partial_method_name === "") {
+                    this.autocomplete_options = [];
```

Comment:
Consider having a minimum length (e.g. 3 characters) before searching for auto complete results

Reply:
I think it's fine - Bob Wallis' search refinement is excellent for giving you popular methods - for example, I know that I can type `c` then it enter and I'll get Cambridge on whatever stage.